#56 no drops in dev-tools and option in morphic properties menu

linusha opened the pull request:
Browser, Object Editor, JS Workspace and Inspector do not accept drops anymore.
The morph menu allows to accept/not accept drops for a whole hierarchy of morphs.

![Tooltip_072](https://user-images.githubusercontent.com/14252419/98812893-9fe56b80-2423-11eb-8347-38917e482370.png)

We briefly discussed a better solution to block drops in all dev-tools and e.g. color-pickers as well. Since we did not come up with a good solution we opted for this, since these tools were the biggest pain points for us.

This closes #56

merryman commented on lively.morphic/morph.js:
```diff
@@ -2236,6 +2236,20 @@ export class Morph {
 
+  doNotAcceptDropsForThisAndSubmorphs() {
+    this.acceptsDrops = false;
+    this.submorphs.forEach((m) => {
```
Interesting, I was expecting this to not only look at the direct children of the morph in question, but the entire submorph hierarchy, so all children reachable from that morph.
I can imagine a bunch of scenarios where I have to manually go through each layer of morphs manually and disable the drops, which kind of defeats the purpose of this "shortcut"? 
For instance, calling this on a list will still leave the items in that list unchanged (and continue to accept drops), since these are wrapped inside a list container.
Then again, it avoids unintended disabling of drops, since a smaller set of morphs is altered.
Thoughts?

linusha replied:
I am not sure if I understand you correctly, but I actually thought we traverse the whole submorph-hierarchy, since we have the recursive call in line 2242?

merryman replied:
Beats me! You are absolutely right. Please use the idiomatic way in lively which is: `this.withAllSubmorphsDo(m => {})`
However then we end up with the other side of the issues I mentioned. For instance opening up a drag and drop game inside a window would break the game.

merryman commented on lively.ide/js/browser/index.js:
```diff
@@ -77,7 +77,10 @@ export default class Browser extends Morph {
-    if (!browser.world()) browser.openInWindow(); 
+    if (!browser.world()) {
+      let window = browser.openInWindow();
+      window.doNotAcceptDropsForThisAndSubmorphs();
```
This probably skips a bunch of containers inside the browser who still continue to accept drops. From the top of my head I can think of the wrapper for the top row of buttons as one example of that.

merryman commented on lively.ide/js/objecteditor/index.js:
```diff
@@ -60,6 +60,7 @@ export class ObjectEditor extends Morph {
         evalEnvironment
       });
     }
+    win.doNotAcceptDropsForThisAndSubmorphs();
```
Same as with the browser, what happens for instance with the importController? It will happily continue accepting drops. So either more calls of this are needed or we need to adjust the semantics of this method. (See last comment)

merryman reviewed (changes requested):
Overall good idea, but I think we need to pour more thoughts into the actual semantics.

linusha commented:
@merryman as we discussed on monday, we rebased and switched to the idiomatic way to traverse the submorph hierarchy.
Therefore I think, if you find the time, that this can be merged :)

linusha commented on lively.morphic/morph.js:
```diff
@@ -2256,6 +2256,14 @@ export class Morph {
   }
 
+  doNotAcceptDropsForThisAndSubmorphs () {
+    this.withAllSubmorphsDo(m => { m.acceptsDrops = false; debugger; });
```
Aargh, super embarrassing. Sorry!
